### PR TITLE
Python bindings

### PR DESCRIPTION
**Pull request by qwattash**
Add python bindings for the streamtrace and disassembler interfaces.

Dependencies:
compile-time: SWIG
tests: pytest

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,7 +48,7 @@ FIND_PACKAGE(LibLZMA REQUIRED)
 
 
 include_directories(SYSTEM ${LibLZMA_INCLUDE_DIRS})
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LLVM_CXXFLAGS} ${LLVM_VERSION}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LLVM_CXXFLAGS} ${LLVM_VERSION} -fno-strict-aliasing")
```
Comment (davidchisnall) [inline CMakeLists.txt]: Did we determine if this is actually needed?  I thought the bug that we thought might be caused by it was actually a qemu bug?

File: streamtrace.hh [hunk 2]
```diff
@@ -229,6 +232,19 @@ struct capability_register
 static_assert(sizeof(capability_register) <= 32,
               "Capability register structure has grown far too big!");
 
+union debug_trace_register
```
Comment (davidchisnall) [inline streamtrace.hh]: Why is this pulled out here?

File: streamtrace.cc [hunk 3]
```diff
@@ -1875,7 +1888,7 @@ void keyframe::update(const debug_trace_entry &e, disassembler::disassembler &di
 		regs.gpr[gpr-1] = e.reg_value.gp;
 		regs.valid_gprs[gpr-1] = true;
 	}
-	if (e.capreg_number() > 0)
+	if (e.capreg_number() > 0) /* XXXAM: should it be >= or we are ignoring $c0? */
```
Comment (davidchisnall) [inline streamtrace.cc]: Yup, that's a bug.

File: streamtrace.cc [hunk 4]
```diff
@@ -1865,6 +1865,19 @@ debug_trace_entry::debug_trace_entry(const debug_trace_entry_disk_v1 &d,
 	decode_entry(*this, d.version, d.val1, d.val2);
 }
 
+debug_trace_entry::debug_trace_entry() :
```
Comment (davidchisnall) [inline streamtrace.cc]: Don't do this.  Making `debug_trace_entry` have an out-of-line non-trivial default constructor will result in a bunch of redundant work being done.  If you must have the fields zero-initialised, then provide inline initialisers in the class definition.

File: Tests/csetbounds.cc [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#include "streamtrace.hh"
```
Comment (davidchisnall) [inline Tests/csetbounds.cc:1]: Please document what this test is checking in the comments.

File: Python/callback.hh [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+ * the SWIG target languages (see SWIG directors).
+ * The c_scanner method is bound to the director object and
+ * given to the trace::scan method, the c_scanner in turns invokes
+ * Scanner::run which should be implemented by the callback class
+ * in the target language.
+ *
+ * XXX: The three proxy classes may be merged in a template
+ */
+
+struct Scanner
```
Comment (davidchisnall) [inline Python/callback.hh]: I'm not sure why you need these.  You should be able to use any class that defines `operator::()` (including a lambda) and you ought to be able to do this in a private namespace for the bindings, rather than defining a new class.

Looking at your bindings code, can't you just pass a lambda to the scan functions that calls back into the Python interpreter with the Python function that you pass?

File: streamtrace.hh (hunk 2 above)
Comment (qwattash) [inline streamtrace.hh]: SWIG does not currently support nested unions. It is needed to make `debug_trace_entry::reg_value` accessible from python.

Comment (davidchisnall) [inline streamtrace.hh]: It would probably be cleaner for Python to add accessors (which would check that the register is the correct type), rather than exposing the union directly.

File: CMakeLists.txt [hunk 7]
```diff
@@ -48,7 +48,7 @@ FIND_PACKAGE(LibLZMA REQUIRED)
 
 
 include_directories(SYSTEM ${LibLZMA_INCLUDE_DIRS})
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LLVM_CXXFLAGS} ${LLVM_VERSION} -fno-strict-aliasing")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LLVM_CXXFLAGS} ${LLVM_VERSION}")
```
Comment (davidchisnall) [inline CMakeLists.txt]: Did you check if this is still needed?  I don't think it is, we just turned it on to check if the qemu bug was caused by it.

File: Python/cheritrace.i [hunk 8]
```diff
@@ -1,80 +1,250 @@
...
+ * Convert from python function to C function pointer for trace::scanner
+ * trace::detailed_scanner, trace::filter and trace::notifier
+ */
+%typemap (in)
+cheri::streamtrace::trace::scanner
+{
+     if (!PyCallable_Check($input)) {
+	  SWIG_exception(SWIG_TypeError, "Object not callable");
+     }
+     $1 = [$input](cheri::streamtrace::debug_trace_entry entry, uint64_t idx) {
```
Comment (davidchisnall) [inline Python/cheritrace.i:52]: What is the type of $input?

File: Python/cheritrace.i [hunk 9]
```diff
@@ -1,80 +1,250 @@
...
-    std::shared_ptr<trace_view> filter_trace(cheri::streamtrace::Filter* filter) {
-      auto bound_cbk = std::bind(&cheri::streamtrace::Filter::c_filter,
-				 filter,
-				 std::placeholders::_1);
-      return $self->filter(cheri::streamtrace::trace::filter_predicate(bound_cbk));
-    }
+
+%extend cheri::streamtrace::debug_trace_entry
+{
+     /* Workaround missing support of nested unions */
```
Comment (davidchisnall) [inline Python/cheritrace.i]: You can make these methods visible in C++ if you like - they're probably better style anyway.

File: Python/cheritrace.i (hunk 8 above)
Comment (qwattash) [inline Python/cheritrace.i:52]: It is a PyObject*, it points to the python callback given as the first argument to the scan function. It is the same in all the other typemaps.

File: Python/cheritrace.i (hunk 9 above)
Comment (qwattash) [inline Python/cheritrace.i]: Ok, I will move these to the trace_entry.

File: CMakeLists.txt [hunk 10]
```diff
@@ -77,6 +77,7 @@ if (TESTS)
 endif ()
 
 add_subdirectory(Tools)
+add_subdirectory(Python)
```
Comment (davidchisnall) [inline CMakeLists.txt]: Please make this conditional on a flag (default to true, but let people turn it off).

File: Tools/tracedump.cc [hunk 11]
```diff
@@ -216,17 +216,20 @@ inline void print_register(const debug_trace_entry &e)
 	}
 }
 
+#ifdef SIGINFO
```
Comment (davidchisnall) [inline Tools/tracedump.cc]: This doesn't need to be conditional on SIGINFO, it will just be dead code.  Mark the variable and function `static` and they'll be optimised away if not used.

File: Tools/tracedump.cc [hunk 12]
```diff
@@ -310,6 +313,7 @@ int main(int argc, char **argv)
 					putchar('\n');
 				}
 			}
+#ifdef SIGINFO
```
Comment (davidchisnall) [inline Tools/tracedump.cc]: Similarly here, don't make it conditional, let the compiler do the DCE.

File: disassembler.cc [hunk 13]
```diff
@@ -77,7 +77,7 @@ int registerIndexForString(const char *str)
 		}
 		return idx + 64;
 	}
-	if (str[0] == 'f')
+	if (str[0] == 'f' && str[1] != 'p')
```
Comment (davidchisnall) [inline disassembler.cc:80]: Please put this in a separate commit.

File: streamtrace.cc [hunk 14]
```diff
@@ -1875,7 +1875,7 @@ void keyframe::update(const debug_trace_entry &e, disassembler::disassembler &di
 		regs.gpr[gpr-1] = e.reg_value.gp;
 		regs.valid_gprs[gpr-1] = true;
 	}
-	if (e.capreg_number() > 0)
+	if (e.capreg_number() >= 0)
```
Comment (davidchisnall) [inline streamtrace.cc:1887]: And this one.  Ideally with a test.

File: streamtrace.hh [hunk 15]
```diff
@@ -42,9 +42,12 @@ namespace cheri
 namespace disassembler
 {
 	class disassembler;
-};
+}
 namespace streamtrace
 {
+#ifndef SIZE_T_MAX
+#define SIZE_T_MAX SIZE_MAX
+#endif
```
Comment (davidchisnall) [inline streamtrace.hh]: Where is this needed for?  We probably should be using `std::limits<size_t>::max` instead anyway...

Comment (qwattash) [inline streamtrace.hh]: This was needed to build on linux, I will change it to use std::limits

File: CMakeLists.txt (hunk 10 above)
Comment (qwattash) [inline CMakeLists.txt]: There was already a flag, I am moving it to the top-level CMakeLists.

File: disassembler.cc (hunk 13 above)
Comment (qwattash) [inline disassembler.cc:80]: This should already be in a separate commit, see 6ca7ef0

Comment (davidchisnall) [inline disassembler.cc:80]: Can you just push these separately (they're simple and obvious bug fixes - it would be nice to have tests for them though...)
